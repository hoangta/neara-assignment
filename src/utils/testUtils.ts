import { MFA } from "../types/mfa";

export const mockStore = () => {
  jest.mock("../stores/mfaStore", () =>
    jest.fn().mockImplementation(() => ({
      mfas: mockMFAs,
      createMFA: jest.fn(),
      switchMFAs: jest.fn(),
    }))
  );
};

export const mockSystemTime = (date: Date) => {
  jest.useFakeTimers("modern");
  jest.setSystemTime(date);
};

export const mockSVGElementAnimate = () => {
  const circleAnimateSpy = jest.fn();
  SVGElement.prototype.animate = circleAnimateSpy;
  return circleAnimateSpy;
};

export const mockMFAs: MFA[] = [
  {
    provider: "Test Provider",
    code: "012345",
    image: "www.testurl.com",
    exp: new Date(2023, 2, 2),
  },
  {
    provider: "Test Provider 2",
    code: "543210",
    image: "www.testurl.com",
    exp: new Date(2023, 2, 3),
  },
];
