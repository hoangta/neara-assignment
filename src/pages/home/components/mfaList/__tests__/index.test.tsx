import React from "react";
import { fireEvent, render } from '@testing-library/react'
import MFAList from '..'
import { mockMFAs, mockSVGElementAnimate } from "../../../../../utils/testUtils";

jest.useFakeTimers("modern");
jest.setSystemTime(new Date(2023, 2, 1));
mockSVGElementAnimate();

describe('Pages/Home/Components/MFAList', () => {
    const switchMFAsSpy = jest.fn();

    it('should render correctly', () => {
        const { asFragment } = render(<MFAList mfas={mockMFAs} isEditing switchMFAs={switchMFAsSpy} />);
        expect(asFragment()).toMatchSnapshot();
    })

    it('should trigger switchMFAs correctly', () => {
        const { getByTestId } = render(<MFAList mfas={mockMFAs} isEditing switchMFAs={switchMFAsSpy} />);
        fireEvent.click(getByTestId('moveup_button_0'));
        expect(switchMFAsSpy).toBeCalledWith(0, -1);
        fireEvent.click(getByTestId('movedown_button_0'));
        expect(switchMFAsSpy).toBeCalledWith(0, 1);
        fireEvent.click(getByTestId('moveup_button_1'));
        expect(switchMFAsSpy).toBeCalledWith(1, 0);
        fireEvent.click(getByTestId('movedown_button_1'));
        expect(switchMFAsSpy).toBeCalledWith(1, 2);
    })
})