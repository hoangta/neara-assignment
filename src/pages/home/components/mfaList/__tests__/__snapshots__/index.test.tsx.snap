// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages/Home/Components/MFAList should render correctly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="container"
    >
      <div
        class="moveButtonContainer"
      >
        <button
          class="moveButton"
          data-testid="moveup_button_0"
        >
          ⬆
        </button>
        <button
          class="moveButton"
          data-testid="movedown_button_0"
        >
          ⬇
        </button>
      </div>
      <img
        alt="Test Provider"
        class="image"
        src="www.testurl.com"
      />
      <div
        class="detail"
      >
        <div
          class="provider"
        >
          Test Provider
        </div>
        <div
          class="code"
        >
          012 345
        </div>
      </div>
      <div
        class="indicatorContainer"
      >
        <svg
          class="indicatorItem"
        >
          <circle
            class="circle"
            cx="20"
            cy="20"
            fill="none"
            r="14"
            stroke="gray"
            stroke-dasharray="87.96459430051421"
            stroke-dashoffset="87.96459430051421"
            stroke-width="2"
          />
        </svg>
        <div
          class="countdownLabel"
        >
          2464028
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="moveButtonContainer"
      >
        <button
          class="moveButton"
          data-testid="moveup_button_1"
        >
          ⬆
        </button>
        <button
          class="moveButton"
          data-testid="movedown_button_1"
        >
          ⬇
        </button>
      </div>
      <img
        alt="Test Provider 2"
        class="image"
        src="www.testurl.com"
      />
      <div
        class="detail"
      >
        <div
          class="provider"
        >
          Test Provider 2
        </div>
        <div
          class="code"
        >
          543 210
        </div>
      </div>
      <div
        class="indicatorContainer"
      >
        <svg
          class="indicatorItem"
        >
          <circle
            class="circle"
            cx="20"
            cy="20"
            fill="none"
            r="14"
            stroke="gray"
            stroke-dasharray="87.96459430051421"
            stroke-dashoffset="87.96459430051421"
            stroke-width="2"
          />
        </svg>
        <div
          class="countdownLabel"
        >
          2550428
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
